File: python_programmers_school/level 1 review/자릿수 더하기.js
// 한줄로 작성하는 방법
function solution(n) {
  const answer = String(n)
    .split("")
    .reduce((sum, cur) => sum + Number(cur), 0);
  return answer;
}

// 풀어쓰기
function solution(n) {
  let answer = 0;
  for (s of String(n)) {
    answer += parseInt(s);
  }
  return answer;
}

// 다른 사람의 풀이
// n을 문자열로 바꾸는 다른 방법
// js에서는 더하기 연산에서 피연산자 중 하나가 문자열이면 숫자가 문자열로 변환되어 계산 결과가 문자열끼리 더한 값이 된다.

function solution(n) {
  return (n + "").split("").reduce((acc, curr) => acc + parseInt(curr), 0);
}

File: python_programmers_school/level 1 review/자릿수 더하기.py
def solution(number):
    answer = 0
    for s in str(number):
        answer += int(s)
    return answer


# 다른 사람의 풀이 1 (map 사용)
def sum_digit(number):
    return sum(map(int, str(number)))     

# 다른 사람의 풀이 2 (재귀 사용)
# 321%10은 123을 10으로 나눈 나머지 3을 반환합니다. 123//10은 123을 10으로 나눈 몫 12를 반환합니다. 따라서 12를 다시 함수에 넣고 
# 돌리겠죠. (return 123%10 + sum_disgit(123//10) -> return 3 + sum_digit(12)).2번째 돌릴땐 return에는 이전에 반환한 3 에다가 
# + 12%10 + sum_digit(12//10). 즉 2번째에는 return 3+2+sum_digit(1)이 됩니다. 3번째에는 if문에서 1이 10보다 작으므로 return 1을 
# 해주고 최종적으로 return 3+2+1이 됩니다.
def sum_digit(number):
    if number < 10:
        return number

    return number%10 + sum_digit(number//10)